DuckDB.NET.Data/DuckDBStatementExecutor.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Collections.Generic;
+
+namespace DuckDB.NET.Data;
+
+internal static class DuckDBStatementExecutor
+{
+    private static readonly Dictionary<Type, Func<DuckDBPreparedStatement, long, object, DuckDBState>> Binders =
+        new Dictionary<Type, Func<DuckDBPreparedStatement, long, object, DuckDBState>>
+        {
+            {typeof(bool), BindBoolean},
+            {typeof(sbyte), BindInt8},
+            {typeof(short), BindInt16},
+            {typeof(int), BindInt32},
+            {typeof(long), BindInt64},
+            {typeof(float), BindFloat},
+            {typeof(double), BindDouble},
+            {typeof(string), BindString},
+        };
+
+    public static DuckDBQueryResult Execute(DuckDBNativeConnection connection, string query, DuckDBDbParameterCollection parameterCollection)
+    {
+        using var unmanagedQuery = query.ToUnmanagedString();
+        var queryResult = new DuckDBResult();
+        DuckDBPreparedStatement preparedStatement = null;
+        try
+        {
+            var result = NativeMethods.PreparedStatements.DuckDBPrepare(connection, unmanagedQuery, out preparedStatement);
+            if (!result.IsSuccess())
+            {
+                var errorMessage = NativeMethods.PreparedStatements.DuckDBPrepareError(preparedStatement).ToManagedString(false);
+                throw new DuckDBException(string.IsNullOrEmpty(errorMessage) ? "DuckDBQuery failed" : errorMessage, result);
+            }
+
+            BindParameters(preparedStatement, parameterCollection);
+            
+            result = NativeMethods.PreparedStatements.DuckDBExecutePrepared(preparedStatement, queryResult);
+            if (!result.IsSuccess())
+            {
+                var errorMessage = NativeMethods.Query.DuckDBResultError(queryResult).ToManagedString(false);
+                throw new DuckDBException(string.IsNullOrEmpty(errorMessage) ? "DuckDBQuery failed" : errorMessage, result);
+            }
+
+            var tmp = queryResult;
+            queryResult = null;
+            return new DuckDBQueryResult(tmp);
+        }
+        finally
+        {
+            preparedStatement?.Dispose();
+            if (queryResult != null)
+                NativeMethods.Query.DuckDBDestroyResult(queryResult);
+        }
+    }
+
+    private static void BindParameters(DuckDBPreparedStatement preparedStatement, DuckDBDbParameterCollection parameterCollection)
+    {
+        if (parameterCollection == null || parameterCollection.Count == 0) return;
```

Comment:
What if the command text has parameters but none of them are specified?

Reply:
That's what I'm saying, I think we should report it, just like you check it on the next lines. I think you need to always do that check, even when `parameterCollection` is empty